File: Python/Substrings of Size Three with Distinct Characters.py
from collections import Counter


class Solution:
    """
    Time:   O(n)
    Memory: O(min(k, n))
    """

    K = 3

    def countGoodSubstrings(self, s: str) -> int:
        window = Counter(s[:self.K])
        count = int(len(window) == self.K)

        for i in range(self.K, len(s)):
            if window[s[i - self.K]] > 1:
                window[s[i - self.K]] -= 1
            else:
                window.pop(s[i - self.K])
            window[s[i]] += 1

            count += int(len(window) == self.K)

        return count

File: Java/Substrings of Size Three with Distinct Characters.java
import java.util.HashMap;

class Solution {

    /**
     * Time:   O(n)
     * Memory: O(min(n,k))
     */
    public int countGoodSubstrings(String s) {
        HashMap<Character, Integer> window = new HashMap<>();

        for (int i = 0; i < Math.min(K, s.length()); ++i)
            window.put(s.charAt(i), window.getOrDefault(s.charAt(i), 0) + 1);

        int count = window.size() == K ? 1 : 0;

        for (int i = K; i < s.length(); ++i) {
            if (window.getOrDefault(s.charAt(i - K), 0) > 1)
                window.put(s.charAt(i - K), window.get(s.charAt(i - K)) - 1);
            else
                window.remove(s.charAt(i - K));

            window.put(s.charAt(i), window.getOrDefault(s.charAt(i), 0) + 1);
            count += window.size() == K ? 1 : 0;
        }

        return count;
    }

    private static final int K = 3;
}
